iib/workers/tasks/utils.py

```diff
@@ -23,6 +24,139 @@
+    resolved_deprecation_list = get_resolved_bundles(deprecation_list)
+    deprecate_bundles = []
+    for bundle in bundles:
+        if bundle in resolved_deprecation_list:
+            deprecate_bundles.append(bundle)
+
+    log.info(
+        'Bundles that will be deprecated from the index image: %s', ', '.join(deprecate_bundles)
+    )
+    return deprecate_bundles
+
+
+def get_resolved_bundles(bundles):
+    """
+    Get the pull specification of the bundle images using their digests.
+
+    Determine if the pull spec refers to a manifest list.
+    If so, simply use the digest of the first item in the manifest list.
+    If not a manifest list, it must be a v2s2 image manifest and should be used as it is.
+
+    :param list bundles: the list of bundle images to be resolved.
+    :return: the list of bundle images resolved to their digests.
+    :rtype: list
+    :raises IIBError: if unable to resolve a bundle image.
+    """
+    log.info('Resolving bundles %s', ', '.join(bundles))
+    resolved_bundles = set()
+    for bundle_pull_spec in bundles:
+        skopeo_raw = skopeo_inspect(f'docker://{bundle_pull_spec}', '--raw')
+        if (
+            skopeo_raw.get('mediaType')
+            == 'application/vnd.docker.distribution.manifest.list.v2+json'
+        ):
+            # Get the digest of the first item in the manifest list
+            digest = skopeo_raw['manifests'][0]['digest']
+            name = _get_container_image_name(bundle_pull_spec)
+            resolved_bundles.add(f'{name}@{digest}')
+        elif (
+            skopeo_raw.get('mediaType') == 'application/vnd.docker.distribution.manifest.v2+json'
+            and skopeo_raw.get('schemaVersion') == 2
+        ):
+            resolved_bundles.add(get_resolved_image(bundle_pull_spec))
+        else:
+            error_msg = (
+                f'The pull specification of {bundle_pull_spec} is neither '
+                f'a v2 manifest list nor a v2s2 manifest. Type {skopeo_raw.get("mediaType")}'
+                f' and schema version {skopeo_raw.get("schemaVersion")} is not supported by IIB.'
+            )
+            raise IIBError(error_msg)
+
+    return list(resolved_bundles)
+
+
+def _get_container_image_name(pull_spec):
+    """
+    Get the container image name from a pull specification.
+
+    :param str pull_spec: the pull spec to analyze
+    :return: the container image name
+    """
+    if '@' in pull_spec:
+        return pull_spec.split('@', 1)[0]
+    else:
+        return pull_spec.rsplit(':', 1)[0]
+
+
+def get_resolved_image(pull_spec):
+    """
+    Get the pull specification of the container image using its digest.
+
+    :param str pull_spec: the pull specification of the container image to resolve
+    :return: the resolved pull specification
+    :rtype: str
+    """
+    log.debug('Resolving %s', pull_spec)
+    name = _get_container_image_name(pull_spec)
+    skopeo_output = skopeo_inspect(f'docker://{pull_spec}', '--raw', return_json=False)
```

Comment:
huh odd, why do we pass `return_json=False` only to use `json.loads` on the next line?

Reply:
/me assumes it's due to a skopeo quirk